File: epi_cpp/1_PrimitiveTypes/15_intersecting_rectangle.cpp
#include <iostream>

using namespace std;

struct Rectangle{
    int x, y, width, height;
};

bool IsIntersect(const Rectangle& R1, const Rectangle& R2){
    return (R1.x <= R2.x + R2.width && R1.x + R1.width >= R2.x) &&
           (R1.y <= R2.y + R2.height && R1.y + R1.height >= R2.y);
}


Rectangle IntersectRectangle(const Rectangle& R1, const Rectangle& R2){
    if(IsIntersect(R1, R2)){
        return {
            max(R1.x, R2.x), max(R1.y, R2.y),
            min(R1.x + R1.width, R2.x + R2.width) - max(R1.x, R2.x),
            min(R1.y + R1.height, R2.y + R2.height) - max(R1.y, R2.y)
        };
    }
    return {0, 0, -1, -1};
}

int main(){
    Rectangle r1 = Rectangle{2, 4, 5, 5};
    Rectangle r2 = Rectangle{3, 5, 5, 5};
    auto result = IntersectRectangle(r1, r2);
    cout << "{"<< result.x <<", "<<result.y<<", "<<result.width <<", "<<result.height<<"}"<<endl;
    return 0;
}

File: epi_python/01_PrimitiveTypes/15_intersecting_rectangle.py
import collections

Rect = collections.namedtuple('Rect', ('x', 'y', 'width', 'height'))

def intersect_rectangle(r1: Rect, r2: Rect)->Rect:
    def is_intersect(r1, r2):
        return (r1.x <= r2.x + r2.width and r1.x + r1.width >= r2.x
                and r1.y <= r2.y + r2.height and r1.y + r1.height >= r2.y)
    if not is_intersect(r1, r2):
        return Rect(0, 0, -1, -1)
    return Rect(
        max(r1.x, r2.x), max(r1.y, r2.y),
        min(r1.x + r1.width, r2.x + r2.width) - max(r1.x, r2.x),
        max(r1.y + r1.height, r2.y + r2.height) - max(r1.y, r2.y)
    ) 

